src/analysis/souffle/dataflow_graph.dl

```diff
@@ -21,18 +21,49 @@
 // A data flow edge.
 .decl edge(src: AccessPath, tgt: AccessPath)
 
+// Use the previous relations to map an edge to a generated access path
+// representing its midpoint.
+.decl edgeToMidpointAccessPath(src: AccessPath, tgt: AccessPath, midpoint: AccessPath)
+
+// An "innerEdge" is an internal concept. When the user requests an edge with an
+// edge fact, we will expand that into innerEdges in one of two ways:
+//
+// 1. The edge is not the subject of a claimNotEdge, in which case an
+// innerEdge directly corresponding to the edge is created.
+//
+// 2. The edge is the subject of a claimNotEdge, in which case three facts
+// are created: a generated midway point AccessPath, an innerEdge going from the
+// src of the original edge to the midway point, and the midway point AccessPath
+// to the tgt. That allows us to effectively eliminate edges that are the
+// subject of a claimNotEdge from the dataflow graph by eliminating all tags
+// from the midway point. Paths are defined in terms of innerEdges.
+.decl innerEdge(src: AccessPath, tgt: AccessPath)
+
 // A direct or transitive data flow path.
 .decl path(src: AccessPath, tgt: AccessPath)
 
 //-----------------------------------------------------------------------------
 // Rules
 //-----------------------------------------------------------------------------
+// Generate a midpoint access path for each edge that is the subject of a
+// claimNotEdge from any principal.
+edgeToMidpointAccessPath(src, tgt, cat("##midpoint_", to_string($))) :-
+  edge(src, tgt), claimNotEdge(_, src, tgt).
+
+// Generate inner edges to and from that midpoint for each edge with a midpoint.
+innerEdge(src, midpoint), innerEdge(midpoint, tgt) :-
+  edgeToMidpointAccessPath(src, tgt, midpoint).
+
+// For edges for which a midpoint was not generated, generate an innerEdge
+// corresponding to the original edge.
+innerEdge(src, tgt) :- edge(src, tgt), !edgeToMidpointAccessPath(src, tgt, _).
```

Comment:
I am not sure what defining a midpoint gets us. Why isn't this simply the following?
```
innerEdge(src, tgt) :- edge(src, tgt), !claimNotEdge(src, tgt).
```

Reply:
So, first, this isn't the "interesting" case for midpoints. This is generating inner edges for the case where there was not a claim that an edge wasn't there, which produces an inner edge exactly corresponding to the original edge requested by the user.

In the example code that you wrote, we emit an `innerEdge` corresponding to an original `edge` fact when there is not a claim that the edge does not exist. The problem with this approach comes down to the principal that claims that the edge is not there and the degree to which their word is universally taken. `claimNotEdge` takes an argument indicating the principal making that claim like so:

```
.decl claimNotEdge(principal: Principal, src: AccessPath, dst: AccessPath)
```

In the above, `innerEdge`s are shared resources that all tags flow across. If we end up not producing the `innerEdge`, this blocks all tags from flowing along the original `edge` as requested by the claim, thus producing behavior as if all principals trusted the claiming `principal`. If we end up producing the `innerEdge`, all tags can then flow along the edge, producing behavior as if all principals distrusted the claiming `principal`.

The property we are trying to get with this PR is that the tag propagation behavior behaves as if the edge indicated by `claimNotEdge` is removed from the dataflow graph only for those tags belonging to principals which trust the `principal` indicated in `claimNotEdge`. The current solution accomplishes that by generating a `downgrades` said by `principal` for every `tag` in the current DL script's universe of tags on all tags flowing through this edge. Each of these produced `downgrades` will be honored only if the `tag` being downgraded can be downgraded by the `principal` attempting the downgrade. The midpoint is just a way to clear tags on an `AccessPath` on the edge without speaking of an `AccessPath` that may be shared by other edges (ie, the `src` or the `dst`).

Reply:
Ah, I understand the proposal now. It seems like `midpoint` relations are where you do the downgrade. Can't we get achieve the same without the `midpoint` relation?

```
saysDowngrades(principal, tgt, tag):
    mayHaveTag(src, tag), edge(src, tgt), claimNotEdge(principal, src, tgt).
```

(I am not sure if this breaks stratification of the datalog relations for the solver.)

Reply:
What @bgogul suggested looks like the answer, and we should probably use it and get rid of the midpoints :). Because it does the downgrade on the output (`tgt`) handle, but uses the input (`src`) handle to figure out which tags to use, does not have the problem of requesting downgrades on other output edges like [this one does](https://github.com/google-research/raksha/pull/145). It also removes another bit of weirdness from #145 which is that #145 would request to remove tags that are not even on the edge.

I mistakenly thought that the downgrade needed to pull tags off of the input to a particle, but pulling them off of the outputs works (I did a check on paper).

Reply:
Ach! No it does break stratification :)

Reply:
This PR now has a solution that regains stratifiability with a loss of precision but not safety

[#145 ](https://github.com/google-research/raksha/pull/145/files)

See also this commit message for an explanation: https://github.com/google-research/raksha/pull/145/commits/1e8496249af9e64d42e11a9dfce322be719f0fb0

See what you think. (I have no preference for which PR we use. I am just using this as a way to send code. We probably do need tests which the other PR lacks).

Reply:
I'm not sure that @bgogul 's suggestion works exactly. Imagine that there are two edges from different sources, `edge1` and `edge2`, each of which ends at `tgt`. Imagine that a dataflow graph without authorization logic would propagate `tag` along both edges. Imagine that there is a claim that `edge1` does not exist that is believed by the owner of `tag`. Using the rule proposed by @bgogul above, `principal` will `saysDowngrades` on `tgt`, clearing `tag` from *any* edge source, thus clearing the `tag` that should have propagated down `edge2`.

This is why I've introduced midpoints, as odd as they are: they allow clearing tag taint on a particular edge that is guaranteed to not interfere with the propagation of taint along other edges.

Reply:
I see. The issue is that we need the ability to make a different downgrade claim for each edge. The midpoints are creating "lookup keys" for each edge, which lets us make additional transformations before combining them. The results of different edges will get combined afterwards to give the desired result. 

An alternative way of achieving the same would be to add `src` to the `saysDowngrades` relation, i.e., `saysDowngrade(principal, src, tgt, tag)`, which is the effectively the same as `saysDowngrade(principal, midpoint, tag)` in the PR. We will also need to change `downgrades` to have `src`, i.e., `downgrades(src, tgt, tag)`. Isn't that much easier to understand?  The downside is that downgrades need to know about edges, which may not be ideal.

In any case, I am fine with us going forward with this PR. The "midpoint" access path seems like a useful notion to keep around as it allows us to do additional transformations along an edge. 

However, we should think of a different name as this is generally tied to another relation (`claimNotEdge`) in this case.

Reply:
Opened https://github.com/google-research/raksha/issues/158 to track this discussion.